airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py

```diff
@@ -0,0 +1,349 @@
+        """
+        Flatten extra_params if it contains pagination information
+        """
+        next_page_token = None  # reset it, pagination data is in extra_params
+        if extra_params:
+            page = extra_params.pop("page", {})
+            extra_params.update(page)
+        return super()._request_params(stream_state, stream_slice, next_page_token, extra_params)
+
+    def send_request(self, **kwargs) -> Optional[requests.Response]:
+
+        if self.reqs_per_hour_limit:
+            if self.is_first_request:
+                self.is_first_request = False
+            else:
+                # we skip this block, if self.reqs_per_hour_limit = 0,
+                # in all other cases wait for X seconds to match API limitations
+                # https://help.mixpanel.com/hc/en-us/articles/115004602563-Rate-Limits-for-Export-API-Endpoints#api-export-endpoint-rate-limits
+                self.logger.info(
+                    f"Sleep for {3600 / self.reqs_per_hour_limit} seconds to match API limitations after reading from {self.name}"
+                )
+                time.sleep(3600 / self.reqs_per_hour_limit)
+
+        return super().send_request(**kwargs)
+
+
+class AnnotationsHttpRequester(MixpanelHttpRequester):
+    def get_url_base(self) -> str:
+        """
+        REGION: url
+        app/projects/{{ project_id }}/annotations
+        """
+        url_base = super().get_url_base()
+        project_id = self.config.get("credentials", {}).get("project_id", "")
+        last_part = f"app/projects/{project_id}/" if project_id else "2.0/"
+        return f"{url_base}{last_part}"
+
+    def get_request_params(
+        self,
+        *,
+        stream_state: Optional[StreamState] = None,
+        stream_slice: Optional[StreamSlice] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> MutableMapping[str, Any]:
+        return {}
+
+
+class FunnelsHttpRequester(MixpanelHttpRequester):
+    def get_request_params(
+        self,
+        *,
+        stream_state: Optional[StreamState] = None,
+        stream_slice: Optional[StreamSlice] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> MutableMapping[str, Any]:
+        params = super().get_request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        params["unit"] = "day"
+        return params
+
+
+class CohortMembersSubstreamPartitionRouter(SubstreamPartitionRouter):
+    def get_request_body_json(
+        self,
+        stream_state: Optional[StreamState] = None,
+        stream_slice: Optional[StreamSlice] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> Mapping[str, Any]:
+        # https://developer.mixpanel.com/reference/engage-query
+        cohort_id = stream_slice["id"]
+        return {"filter_by_cohort": f'{{"id":{cohort_id}}}'}
+
+
+class EngageTransformation(RecordTransformation):
+    def transform(
```

Comment:
can you make sure there are issues documenting the need for this component? It would be great to get back to this at a later point and backport the required features to the CDK

Reply:
I will add it to lour list of custom components